Adds details about system macros and special forms (#342)
amazon-ion/ion-docs: _books/ion-1-1/src/macros/special_forms.md

## Special Forms

> When a template-expression is syntactically an S-expression, its first element must be a symbol that matches either a 

Reviewer: There's an empty (`TODO`) section in `defining_macros.md` that aims to introduce "TDL" as a well-defined name that we can reference later. 
![image](https://github.com/user-attachments/assets/86e3b43a-76ab-4b87-8448-2b59f075fc07)
We should figure out the best way to say "expression that can appear in TDL" (template expression? TDL expression?) and use it consistently throughout. For now, could you link to the empty TDL section?

Author: TDL clause?

Reviewer: I think I prefer "TDL expression." To me, "clause" implies a known/expected section of something but the template is basically free-form.

set of keywords denoting the special forms, or the name of a previously-defined macro.
The interpretation of the S-expression’s remaining elements depends on how the symbol resolves.
In the case of macro invocations, the elements following the operator are arbitrary template expressions, but for special forms that is not always the case.

Special forms are "special" precisely because they cannot be expressed as macros and must therefore receive bespoke syntactic treatment.
Since the elements of macro-invocation expressions are themselves expressions, when you want something to not be evaluated that way, it must be a special form.

> Finally, these special forms are part of the template language itself, and are not visible to encoded data;

Reviewer: ```suggestion
Finally, these special forms are part of the template language itself, and are not addressable outside of TDL;
```

the E-expression `(:literal foo)` must necessarily refer to some user-defined macro named literal, not to this special form.

> [!TODO]
> Many of these could be system macros instead of special forms. Being unrepresentable in TDL is not a reason for something
> to be a special form.
> Candidates to be moved to system macros are `if_*` and `fail`.
> Additionally, the system macro `parse_ion` may need to be classified as a special form since it only accepts literals.

### `literal`

```ion
(macro USD_price (dollars) (.price dollars (.literal USD)))
```

> In this template, we cannot write `(.price dollars USD)` because the symbol `USD` would be treated as an unbound variable reference and a syntax error, so we turn it into literal data by "escaping" it with `literal`. As an aside, there is no need for such a form in E-expressions, because in that context symbols and S-expressions are not "evaluated", and everything is literal except for E-expressions (which are not data, but encoding artifacts).

Reviewer: The aside might make a nice `> [!TIP]`?

### `if_none`

```ion
(macro if_none (stream* true_branch* false_branch*) /* Not representable in TDL */)
```

The `if_none` form is if/then/else syntax testing stream emptiness.
It has three sub-expressions, the first being a stream to check. 

> If and only if that stream is empty (it produces no values), the second sub-expression is expanded and its results are returned by the `if_none` expression. Otherwise, the third sub-expression is expanded and returned.

Reviewer: There are several references to "returning" values; I think it's somewhat more correct to say "expands to" or "produces" in the context of macros, though I don't feel very strongly about it.

> [!Note]

> > Exactly one branch is expanded, because otherwise the empty `stream` might be used in a context that requires a value, resulting in an errant expansion error.

Reviewer: ```suggestion
> Exactly one branch is expanded, because otherwise the empty `stream` might be used in a context that requires a value, resulting in an errant expansion.
```
or
```suggestion
> Exactly one branch is expanded, because otherwise the empty `stream` might be used in a context that requires a value, resulting in an expansion error.
```
?

```ion

> (macro temperature (degrees scale) {degrees: degrees, scale: (.if_none scale (.literal K) scale)})

Reviewer: Subjectively, I think it helps readability to have non-trivial template expressions pretty-printed starting on the next line.

```
```ion
(:temperature 96 F)     ⇒ {degrees:96,  scale:F}
(:temperature 283 (::)) ⇒ {degrees:283, scale:K}
```

To refine things a bit further, trailing voidable arguments can be omitted entirely:
```ion
(:temperature 283) ⇒ {degrees:283, scale:K}
```

> [!TIP]

> > You can define a macro that wraps `if_none` to create a void-coalescing operator.

Reviewer: s/`void`/`none`/ ?

> ```ion

> > (macro coalesce (maybe_none* else+) (.if_none maybe_none else maybe_none))

Reviewer: I tripped on the name `else`; maybe `default_expr`?

```

```ion
(:foo (::))     => { foo: null }
(:foo 2)        => { foo: [2] }
(:foo (:: 2 3)) => { foo: [2, 3] }
```

The `false_branch` parameter may be elided, allowing `if_some` to serve as a _map-if-not-void_ function.

Example:
```ion
(macro foo ($foo)
       {
         foo: (.if_some $foo [$foo])
       })
```

```ion
(:foo (::))     => { }
(:foo 2)        => { foo: [2] }
(:foo (:: 2 3)) => { foo: [2, 3] }
```

### `if_single`

```ion
(macro if_single (expressions* true_branch* false_branch*) /* Not representable in TDL */)
```

If `expressions` evaluates to exactly one value, it returns `true_branch`. Otherwise, it returns `false_branch`.
Exactly one branch is evaluated. The `true_branch` and `false_branch` arguments are only evaluated if they are to be returned.

### `if_multi`

```ion
(macro if_multi (expressions* true_branch* false_branch*) /* Not representable in TDL */)
```

If `expressions` evaluates to more than one value, it produces `true_branch`. Otherwise, it produces `false_branch`.
Exactly one branch is evaluated. The `true_branch` and `false_branch` arguments are only evaluated if they are to be returned.

### `for`

```ion
(for name_and_values template)
```

> `name_and_values` is a list or s-expression containing one or more s-expressions of the form `(name value1 value2 ... valueN)`. The first value is a symbol to act as a variable name. The remaining values in the s-expression will be visited one at a time; for each value, expansion will produce a copy of the `template` argument expression with any appearance of the variable name replaced by the value.

Reviewer: We'll need to revisit this if we decide to use `(%var)` syntax.

I think `(name value1 value2 ... valueN)` should be `(name expr1 expr2 ... exprN)` and an explanation that the `exprN` streams' output will be concatenated à la `.values`.

For example:

```ion

> (:for

Reviewer: Here and below, I think this should be `.for` since you can't use the special form as an e-expression.

```

If `expressions` evaluates to more than one value, it produces `true_branch`. Otherwise, it produces `false_branch`.
Exactly one of `true_branch` and `false_branch` is evaluated.
The `stream` expression must be expanded enough to determine whether it produces more than one value, but implementations are not required to fully expand the expression.

### `for`

```ion
(for name_and_expressions template)
```

> `name_and_values` is a list or s-expression containing one or more s-expressions of the form `(name expr0 expr1 ... exprN)`.

Reviewer: ```suggestion
`name_and_expressions` is a list or s-expression containing one or more s-expressions of the form `(name expr0 expr1 ... exprN)`.
```
